Remove trailing slash from void HTML elements

ahukkanen opened the pull request:
### Motivation / Background

This Pull Request has been created because W3C HTML validator gives a lot of the following type of notices about the trailing slashes on void HTML elements:

> Trailing slash on void elements [has no effect](https://github.com/validator/validator/wiki/Markup-%C2%BB-Void-elements#trailing-slashes-in-void-element-start-tags-do-not-mark-the-start-tags-as-self-closing) and [interacts badly with unquoted attribute values](https://github.com/validator/validator/wiki/Markup-%C2%BB-Void-elements#trailing-slashes-directly-preceded-by-unquoted-attribute-values).

### Detail

This Pull Request changes the trailing slash ` />` on void HTML elements to `>` as suggested by the W3C validator.

This changes also the default value of the `tag` helper's `open` argument from `false` to `nil` and reflects this change in the documentation of the method.

The trailing slashes are removed from all void elements but preserved for the self-closing elements within SVGs, i.e. `animate`, `animateMotion`, `animateTransform`, `circle`, `ellipse`, `line`, `path`, `polygon`, `polyline`, `rect`, `set`, `stop`, `use` and `view` as defined here:

https://github.com/rails/rails/blob/185157fe18f33349536ebeddfc792974d81ce762/actionview/lib/action_view/helpers/tag_helper.rb#L112-L125

This is to prevent any SVGs breaking due to this change.

In addition, a new configuration value `config.action_view.void_element_trailing_slash` is added that can be used to revert to the legacy behavior, i.e. keeping the trailing slashes on the void elements. This is to help migrating to newer versions and to provide this as an option in case some implementations rely on the current behavior.

### Additional information

Fix #51643

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

ahukkanen committed:
Remove trailing slash from void HTML elements

ahukkanen committed:
Ensure proper handling of the `open` argument regardless of config

Always respect the `open` argument regardless of the value of the
`void_element_trailing_slash` configuration.

ahukkanen committed:
Fix broken tests for the TagHelper

ahukkanen committed:
Test the trailing slash on void elements

ahukkanen committed:
Add CHANGELOG entry

ahukkanen committed:
Fix more broken tests

rafaelfranca commented on actionview/CHANGELOG.md:
```diff
@@ -94,4 +94,26 @@
 
     *Akhil G Krishnan*
 
+*   Removed the trailing slash from void HTML elements, except for elements included inside SVG tags.
```
Those should be in the top of the file

ahukkanen replied:
Moved to the top by 4b11ace.

rafaelfranca reviewed (changes requested):
Thank you for the pull request. There are a few things that need to change in order for this to be merged. Can you take a look?

skipkayhil commented on actionview/lib/action_view/helpers/tag_helper.rb:
```diff
@@ -245,7 +247,7 @@ def tag_string(name, content = nil, options, escape: true, &block)
         end
 
-        def self_closing_tag_string(name, options, escape = true, tag_suffix = " />")
+        def self_closing_tag_string(name, options, escape = true, tag_suffix = (void_element_trailing_slash == false ? ">" : " />"))
```
I'm not sure I understand why this change is being made. Is this method being used somewhere outside of `define_self_closing_element` and `define_void_element`? The idea in 343897980cf9c1a8c1d102df3afb0631f5c9145b was that this method is for self closing elements but can be specialized for void elements by passing in the suffix.

Since omitting the `/` is _not_ valid for self-closing elements, I'm not sure it should ever be the default for this method.

ahukkanen replied:
This method is used for both: void elements and the self closing SVG tags.

I moved the configuration check to the void element definition method and changed the default value to the tag closing without the ending slash at dcf7b87. This is because going forward, the ending slash should be rather an exception (i.e. SVGs) than the default.

skipkayhil replied:
> This is because going forward, the ending slash should be rather an exception (i.e. SVGs) than the default.

Right, I don't think this is a good idea. Void elements are a special case of self-closing tags where the `/` is optional instead of required. If the default is `>` then every caller wanting a self-closing tag _must_ pass in `/>` to get a valid tag. If the default is `/>` then passing in `>` is just an optimization for void tags but the method won't generate invalid HTML.

ahukkanen replied:
The default can be changed back as per this request but I'll wait for the answer to the other question before doing that.

ahukkanen committed:
Refert the rails-ujs JS files

ahukkanen committed:
Move the CHANGELOG item as first in the list

ahukkanen committed:
Ensure text helper line breaks are respecting the trailing slash

ahukkanen committed:
Ensure the utf8_enforcer_tag is respecting the trailing slash

ahukkanen committed:
Default the trailing slash config to `true` for older configurations

- When using `load_defaults "7.1"` or below, default the new
  configuration to `true`
- When using `load_defaults "7.2"` or higher, default the new
  configuration to `false`

ahukkanen committed:
Add documentation for `config.action_view.void_element_trailing_slash`

ahukkanen committed:
Correct the tag suffix for void elements generated by TagBuilder

ahukkanen commented on actionview/app/assets/javascripts/rails-ujs.esm.js:
Reverted by ee8e8d5.

ahukkanen commented on actionview/app/assets/javascripts/rails-ujs.js:
Reverted by ee8e8d5.

ahukkanen commented on actionview/lib/action_view/helpers/text_helper.rb:
```diff
@@ -533,7 +533,7 @@ def split_paragraphs(text)
           return [] if text.blank?
 
           text.to_str.gsub(/\r\n?/, "\n").split(/\n\n+/).map! do |t|
-            t.gsub!(/([^\n]\n)(?=[^\n])/, '\1<br />') || t
```
Fixed by 5be711e.

ahukkanen commented on actionview/lib/action_view/helpers/form_tag_helper.rb:
```diff
@@ -907,7 +907,7 @@ def utf8_enforcer_tag
         # needs to be the right character regardless of which encoding the
         # browser infers.
-        '<input name="utf8" type="hidden" value="&#x2713;" autocomplete="off" />'.html_safe
+        '<input name="utf8" type="hidden" value="&#x2713;" autocomplete="off">'.html_safe
```
Fixed by d2b41e3.

ahukkanen commented on actionview/lib/action_view/railtie.rb:
```diff
@@ -14,6 +14,7 @@ class Railtie < Rails::Engine # :nodoc:
     config.action_view.image_decoding = nil
     config.action_view.apply_stylesheet_media_default = true
     config.action_view.prepend_content_exfiltration_prevention = false
+    config.action_view.void_element_trailing_slash = false
```
Fixed by fa7bb74. Documentation added by bc76c55.

skipkayhil replied:
The value here should be `true`, and the `load_default` setting the value to `true` should be removed

ahukkanen replied:
This is changed by 1a47028.

ahukkanen committed:
Fix the versioned default values docs

ahukkanen committed:
Add the new config to the new framework defaults template

ahukkanen commented:
@rafaelfranca @skipkayhil Thank you both for your kind reviews and valuable feedback.

The issues pointed out in the reviews are now resolved.

seanpdoyle commented on actionview/lib/action_view/helpers/tag_helper.rb:
```diff
@@ -72,7 +74,8 @@ def #{method_name}(content = nil, escape: true, **options, &block)
                   tag_string("#{name}", content, options, escape: escape, &block)
                 else
-                  self_closing_tag_string("#{name}", options, escape, ">")
+                  tag_suffix = ActionView::Helpers::TagHelper.void_element_trailing_slash ? " />" : ">"
+                  self_closing_tag_string("#{name}", options, escape, tag_suffix)
```
Are there negative performance impacts from introducing this conditional into each individual call?

Is it possible to extract the suffix String definition so that it's outside the `batch.push(<<~RUBY)` string, then interpolated _back into_ the method String?

ahukkanen replied:
@skipkayhil You are correct that it changes current behavior.

But without this change, e.g. these elements won't have the closing slash even if `config.action_view.void_element_trailing_slash` is set to `true`:
![Void elements without closing slash](https://github.com/rails/rails/assets/864340/a5f9f638-325c-4e07-9d92-7e2c082cb94a)

This source is generated by following the getting started guide of Rails.

It is fine by me to revert the change but then we are not respecting the configuration option in all cases, which was part of the fixes implemented after the previous review.

It's just a question what we want here:

1. Keeping the current behavior of having mixed void HTML tags (what is requested here)
2. Respecting the new configuration option in all cases (what was requested in the review)

seanpdoyle replied:
Depending on how https://github.com/rails/rails/pull/51646/files#r1578010728 is resolved, could the `tag_suffix` definition be inlined into the interpolation?

Then, that value could be interpolated directly into the evaluated Ruby method:

```suggestion
                  
                  self_closing_tag_string("#{name}", options, escape, "#{
                    ActionView::Helpers::TagHelper.void_element_trailing_slash ? " />" : ">"
                  }")
```

seanpdoyle replied:
Furthermore, is there a reason that the `void_element_trailing_slash` attribute need to be defined on `ActionView::Helpers::TagHelper`? Could it be defined on the `ActionView::Helpers::TagHelper::TagBuilder` class directly?

If the `mattr_accessor :void_element_trailing_slash, default: false` line were moved to the `TagBuilder` class, that configuration could be accessed directly from the class method, instead of as a fully-qualified class variable:

```suggestion
                  self_closing_tag_string("#{name}", options, escape, "#{void_element_trailing_slash ? " />" : ">"}")
```

ahukkanen replied:
> Depending on how https://github.com/rails/rails/pull/51646/files#r1578010728 is resolved, could the tag_suffix definition be inlined into the interpolation?

Yes this is of course possible. But I'll just stress that it has no implication on performance at least to my knowledge.

I can do this change if it is insisted but I will wait for the feedback regarding the question on how to proceed with this, as there is the question above which is the desired direction with this.

> Furthermore, is there a reason that the void_element_trailing_slash attribute need to be defined on ActionView::Helpers::TagHelper? Could it be defined on the ActionView::Helpers::TagHelper::TagBuilder class directly?

It is also needed at `TagHelper#tag` method.

I don't have a preference where the configuration option is defined, I just thought it would be natural at the lowest nesting level where it is needed.

ahukkanen replied:
@skipkayhil @rafaelfranca Could you chime in which would be the preferred direction forward, so that we could finalize this.

So there are two alternatives, as pointed out above:

1. Not to change this method which would make the HTML inconsistent regarding the introduced configuration flag, i.e. when the ending slashes are preferred, the HTML would be inconsistent and not respect the configuration option in all places.
2. Following the current direction where the configuration option is respected everywhere, but implement the changes requested here

Which one both of you think is the desired approach?

ahukkanen committed:
Apply review recommendation regarding default configuration values
